woocommerce/woocommerce-gutenberg-products-block
# show discount badge in line items in cart

Fixes #1745

This PR adds product `prices` data to the cart items API. This is used in the client to calculate any effective discount, by comparing `regular_price * quantity` and `line_total`. Note this changes how discount calculation works in client - previous implementation assumed that `line_subtotal` was exclusive of all discounts (turns out this was incorrect assumption).

~~This PR changes what value is returned by the cart items API for `line_subtotal`. This is documented in the schema as the full undiscounted price, and is used on the client to render the `SAVE $101!` badge.~~

~~I couldn't find this data in the cart_item object provided by core, so I'm manually calculating it using the line item product `regular_price` and quantity. This feels a little like a hack – is there a better way to pull this from the core data structure? Are the situations where this logic will break, or where we need to allow for hooks/filters to run?~~

~~Related to that, the API field `line_subtotal` now has a different meaning to the cart_item field `line_subtotal`. Should we use a different name for the API field to avoid confusion, or do we need to revisit how we represent discounts in the API?~~

Note that there is a discount for the [whole cart in current woo core](https://docs.woocommerce.com/wc-apidocs/class-WC_Legacy_Cart.html#_get_cart_discount_total) - but we need a discount or full price for each line item.

### Screenshots
<img width="561" alt="Screen Shot 2020-02-21 at 3 22 37 PM" src="https://user-images.githubusercontent.com/4167300/74999054-6cdb0080-54bf-11ea-86a3-f7175e71b98d.png">

### How to test the changes in this Pull Request:

1. Set up some discounted products in your store. There are lots of ways for discounts to apply (e.g. sale price, or coupons), so try a few combinations :)
2. Add some discounted and undiscounted products to your cart.
3. View cart block (add to a page and publish if needed).
4. Confirm all line item details are correct, in particular prices and discount badge.
